전남대 BE_박진화 4주차 과제 (2단계)

Jinhwa-Park opened the pull request:
Option 관련 기능 추가했습니다.

Jinhwa-Park committed:
feat: migrate prev project codes

Jinhwa-Park committed:
docs: delete unnecessary lines

Jinhwa-Park committed:
docs: add README

Jinhwa-Park committed:
docs: modify typo

Jinhwa-Park committed:
rename: move prev codes to oldTestCodes directory

Jinhwa-Park committed:
feat: add codes about Category

Jinhwa-Park committed:
feat: changed feature about set to update

Jinhwa-Park committed:
refactor: modify ProductController as modified Product entity

Jinhwa-Park committed:
feat: add getEmailFromRequest method to JwtUtil

Jinhwa-Park committed:
refactor: update controllers with clear return types and improve readability

Jinhwa-Park committed:
feat: add new fields(color, imageUrl, description)  and update Category entity

Jinhwa-Park committed:
feat: add path variable support for category deletion

Jinhwa-Park committed:
feat: integrate category handling in ProductController

Jinhwa-Park committed:
feat: update Product entity to include category handling

Jinhwa-Park committed:
feat: update ProductRequest, ProductResponse, ProductService to include category handling

Jinhwa-Park committed:
feat: Add code to WishController to verify ProductId

Jinhwa-Park committed:
"remove: delete unnecessary product.html"

Jinhwa-Park committed:
Merge branch 'jinhwa-park' into step1

Jinhwa-Park committed:
docs: update README

Jinhwa-Park committed:
feat: make addOption page for admin

Jinhwa-Park committed:
refactor: change dependency injection from field to constructor

Jinhwa-Park committed:
feat: implement option features

Jinhwa-Park committed:
feat: add option handle feature

Jinhwa-Park committed:
refactor: change dependency injection from field to constructor (JwtAuthenticationFilter, WebConfig, WishService)

Jinhwa-Park committed:
feat: add option schema

Jinhwa-Park committed:
feat: add redundancy option check feature

Jinhwa-Park committed:
feat: add properties

Jinhwa-Park committed:
refactor: change annotation from korean to english

Jinhwa-Park committed:
Revert "feat: add redundancy option check feature"

This reverts commit b74f7ad6399e7319146f39a3a94be2e2992cea06.

Jinhwa-Park committed:
Revert "Revert "feat: add redundancy option check feature""

This reverts commit cd2982f3e76927c6427344062c53dfb5542342b8.

Jinhwa-Park committed:
feat: add enforcing product have at least one option

Jinhwa-Park committed:
feat: add feat ProductController, OptionDTO, ProductRequest

ProductController: when adding Product, check Option list is empty or
not, using ProductValidator
OptionDTO: add annotation checking validation
ProductRequest: make method possible to include multiple options and to
access option list as adding setOption methods.

pkeugine commented on src/main/java/gift/controller/MemberController.java:
```diff
@@ -40,29 +44,16 @@ public ResponseEntity<Map<String, String>> login(@RequestBody Member loginReques
             response.put("token", token);
             return ResponseEntity.ok(response);
         }
-        return  ResponseEntity.status(401).body(null);
+        return ResponseEntity.status(401).body(Map.of("message", "Invalid credentials"));
```
GlobalExceptionHandler 가 처리하도록 로직을 수정해보면 좋을 것 같아요!

Jinhwa-Park replied:
듣고보니 이런 에러 처리를 GlobalExceptionHandler에 이식하는 게 훨씬 깔끔한 코드가 될 것 같네요! 다른 코드들도 살펴보고 옮길 것들을 생각해보겠습니다.

pkeugine commented on src/main/java/gift/dto/OptionDTO.java:
```diff
@@ -0,0 +1,32 @@
+
+import jakarta.validation.constraints.*;
+
+public class OptionDTO {
```
일단 google java convention 을 한 번 확인해보고, 이름 작성 방식을 바꿔보시면 좋을 것 같아요.
그리고 dto 중 어떤 것들은 `request`, `response` 라는 이름을 사용하시는데,
이번에는 `DTO` 라는 이름을 붙여주셨네요!
이렇게 만들어주신 이유가 궁금합니다.

Jinhwa-Park replied:
request, response를 쓰게 됐던 이유가 다른 분들의 코드를 보고 오 이거 깔끔하게 잘 됐다 싶어서 흉내내봐야지 하고 넣었던 경우인데, convention을 염두에 두고 있지는 않았던 것 같습니다. 일관성을 위해서 OptionDTO는 OptionRequest로 변경하고 convention을 읽고 왜 request, response를 쓰는지를 더 공부해보겠습니다.

pkeugine commented on src/main/java/gift/dto/ProductRequest.java:
```diff
@@ -40,4 +48,15 @@ public String getImageUrl() {
+    public List<OptionDTO> getOptions() {
+        return options;
+    }
+    public void setOptions(List<OptionDTO> options) {
+        this.options = options;
+    }
```
setter 는 이제 슬슬 안만들 때가 된 것 같아요!

Jinhwa-Park replied:
setter를 안만드는 것을 추천하신 이유가 궁금해서 찾아보았습니다.

1. setter 메소드 사용 시 값을 변경한 의도를 파악하기 힘듦
2. 객체의 일관성을 유지하기 어려움.
생성자 오버로딩, Builder 패턴 사용, 정적 팩토리 메소드를 사용하는 것이 좀 더 좋다!

이런 결과를 얻게 되었습니다. 추천하신 이유가 이것 말고 다른 게 더 있는지 궁금합니다!

pkeugine replied:
네 맞습니다.
그러니 제거해보죠!

Jinhwa-Park replied:
builder로 변경했습니다

pkeugine commented on src/main/java/gift/entity/Category.java:
```diff
@@ -0,0 +1,46 @@
+
+    public String getDescription() {return description;}
+
+    public void update(String name, String color,  String imageUrl, String description) {
+        this.name = name;
+        this.description = description;
+        this.imageUrl = imageUrl;
+        this.color = color;
+    }
```
이왕이면 `update` 라는 메서드가 도메인의 중요한 기능중 하나니까 getter 보다 위에 위치하면 어떨까요?
메서드의 순서를 한 번 스스로 정해보고, 이를 본인만의 컨벤션으로 만들어보아요.

Jinhwa-Park replied:
새로 추가한 사항은 왠만하면 가장 하단에 위치하게 하는 습관이 있었는데 메서드의 순서를 고민해보는 것은 좋은 것 같아요! 고민해보겠습니다!

pkeugine commented on src/main/java/gift/entity/Option.java:
```diff
@@ -0,0 +1,30 @@
+        this.quantity = quantity;
+    }
+
+    // (selectable feature) admin only has right to add option.
```
불필요한 주석은 제거해주세요

Jinhwa-Park replied:
admin 관련 기능을 나중에 구현해야겠다는 것을 잊지 않기 위해 남겼던 주석입니다. PR을 보내기 전에 지우는 게 더 좋았을 것 같네요..!

pkeugine commented on src/main/java/gift/entity/Product.java:
```diff
@@ -1,14 +1,12 @@
-import jakarta.persistence.Entity;
-import jakarta.persistence.GeneratedValue;
-import jakarta.persistence.Id;
+import jakarta.persistence.*;
```
wildcard 로 바꿔주신 이유가 궁금해요!
자동으로 바뀐거라면 이를 방지하는 방법을 알아보는 것도 좋을 것 같네요.

Jinhwa-Park replied:
jakarta.persistence 관련이 많아서 깔끔해보이려고 변경했던 것으로 기억합니다. 하지만 같은 이름을 가진 다른 패키지의 클래스가 존재할 때나 코드를 읽을 때 명확하게 어떤 클래스를 import하는 건지 알기 위해서 다시 롤백하는 게 더 나을 것 같네요! 변경하도록 하겠습니다.

pkeugine commented on src/main/java/gift/entity/Product.java:
```diff
@@ -28,44 +26,58 @@ public class Product {
+   @ManyToOne
+   private Category category;
+
+   @OneToMany(mappedBy = "product", cascade = CascadeType.ALL, orphanRemoval = true)
+   private List<Option> options;
+
+    public Product() {
+
+    }
```
이렇게 인덴트가 살짝 다른건 어떻게 발생하는 현상일까요

Jinhwa-Park replied:
위에 있는 것을 지우고 delete로 한 자씩 줄이다가 발생했던 것 같습니다. 이렇게 사용하는 것을 지양해보겠습니다.

pkeugine commented on src/main/java/gift/service/OptionService.java:
```diff
@@ -0,0 +1,48 @@
+    private final OptionRepository optionRepository;
+    private final ProductRepository productRepository;
+
+    @Autowired
+    public OptionService(OptionRepository optionRepository, ProductRepository productRepository) {
+        this.optionRepository = optionRepository;
+        this.productRepository = productRepository;
+    }
```
`@Autowired` 관련 지식을 이번에는 본인의 것으로 만들어보죠.

Jinhwa-Park replied:
넵 좀 더 공부해보겠습니다!

pkeugine replied:
여러번 코멘트를 달았던 것으로 기억해요!

pkeugine commented on src/main/java/gift/service/OptionService.java:
```diff
@@ -0,0 +1,48 @@
+            throw new RuntimeException("Duplicate option name");
+        }
+
+        Option option = new Option();
+        option.update(optionDTO.getName(), optionDTO.getQuantity());
```
생성자를 이용하여 초기화를 해볼 수 있지 않을까요?

Jinhwa-Park replied:
그게 더 깔끔할 것 같습니다. 변경하겠습니다!

pkeugine commented on src/main/java/gift/service/OptionService.java:
```diff
@@ -0,0 +1,48 @@
+        option.update(optionDTO.getName(), optionDTO.getQuantity());
+        optionRepository.save(option);
+
+        product.getOptions().add(option);
```
`getOptions` 를 하지 않고 product 클래스에 책임을 위임해볼까요?

Jinhwa-Park replied:
사실 그 부분을 생각해보긴 했는데 option 관련 코드는 그래도 option 관련 클래스에 있는게 좋지 않을까? 하는 생각으로 그대로 OptionService에 작성했던 기억이 납니다. 단일 책임 원칙을 잘못 이해한 것 같습니다. 그렇게 수정하겠습니다!

pkeugine commented on src/main/java/gift/service/OptionService.java:
```diff
@@ -0,0 +1,48 @@
+    public Option addOptionToProduct(Long productId, OptionDTO optionDTO) {
+        Product product = productRepository.findById(productId)
+                .orElseThrow(() -> new RuntimeException("Product not found"));
+        // redundancy option check
+        Optional<Option> existingOption = optionRepository.findByProductIdAndName(productId, optionDTO.getName());
+        if (existingOption.isPresent()) {
+            throw new RuntimeException("Duplicate option name");
+        }
```
그냥 혹시나해서 남기는 질문인데,
코드를 진화님이 직접 작성하시는걸까요?

Jinhwa-Park replied:
구글링해서 나온 코드를 프로젝트 맞게 변경해서 타이핑하고 있습니다. 이것도 전부 영어로 작성해서 의문을 갖게 되신 거라면 이전 PR 리뷰에 댓글로 답변한 것처럼 중간에 한영키로 변경하는 것을 개인적으로 싫어해서 영어로 주석/메시지를 작성하고 있습니다.

pkeugine commented on src/main/java/gift/dto/ProductRequest.java:
```diff
@@ -40,4 +48,15 @@ public String getImageUrl() {
         this.imageUrl = imageUrl;
     }
+
+    public Long getCategoryId() {return categoryId;}
+
+    public void setCategoryId(Long categoryId) {this.categoryId = categoryId;}
```
위의 코멘트를 남긴 이유는,
붙여넣기 할 때 보이는 현상이나, 신경을 많이 쓰지 않아서 발생하는 현상이 종종 보여서입니다.

Jinhwa-Park replied:
붙여넣기는 똑같은 프로젝트를 하나 만들어서 써보고 그걸 다시 가져오고 있어서 발생한 현상 같습니다.
신경을 많이 쓰지 않은 것은 시정하겠습니다!

pkeugine commented on src/main/resources/schema.sql:
```diff
@@ -18,3 +18,11 @@ CREATE TABLE IF NOT EXISTS wishlist (
+    quantity BIGINT NOT NULL,
+    product_id BIGINT NOT NULL,
+    FOREIGN KEY (product_id) REFERENCES product(id)
+);
```
파일 끝에 빈줄이 없는 현상이 발생하고 있네요.
이 내용은 첫째주에 남겼던 리뷰였던 것 같습니다!

Jinhwa-Park replied:
해당 옵션을 추가한 상태인데 왜 그렇게 됐는지 잘 모르겠습니다. 🤔 
![image](https://github.com/user-attachments/assets/3f263b2e-14f9-44d1-b965-e0c6a48990a2)

pkeugine replied:
앞으론 잘 될 것 같네요! 👍

pkeugine reviewed (changes requested):
안녕하세요 진화님,
step 1 이 아직 머지되지 않았기도 했고,
해당 step 에서도 한 번 확인해주시면 좋을 부분이 있어서 일단 request changes 할게요.
리뷰 확인해주세요!

Jinhwa-Park committed:
fix: make GlobalExceptionHandler handles proper error

Jinhwa-Park committed:
refactor: changes id variable to indicate clearly

Jinhwa-Park committed:
feat: Apply review feedback and add test code

Jinhwa-Park committed:
Merge branch 'step1' into step2

Jinhwa-Park committed:
feat: add builder, delete unnecessary import and annotations

Jinhwa-Park committed:
fix: modify as it works on test codes, change order, add option on data

Jinhwa-Park committed:
feat: add CustomException, delete unnecessary annotations

Jinhwa-Park committed:
rename: modify OptionDTO to OptionRequest

Jinhwa-Park committed:
refactor: Improve validation, error handling, entity retrieval

- Removed unnecessary import or annotation
- added validation error handling with bindingResult
- changed error handling to use CustomException
- Replaced Optional with entity retrieval

Jinhwa-Park committed:
refactor: change to use CustomException

Jinhwa-Park committed:
refactor: change to use CustomException

Jinhwa-Park committed:
refactor: using SpringBootTest to WebMvcTest

Jinhwa-Park committed:
refactor: apply renamed OptionRequest

Jinhwa-Park committed:
refactor: enhance validation

Jinhwa-Park committed:
fix: resolve conflicts

pkeugine reviewed (approved):
해당 단계 이만 머지하겠습니다.
